wgpu/_classes.py

```diff
@@ -701,7 +701,7 @@ class GPUDevice(GPUObjectBase):
     """
 
     def __init__(self, label, internal, adapter, features, limits, queue):
-        super().__init__(label, internal, None)
+        super().__init__(label, internal, self)
```

Comment:
I like the added test to make sure all objects have a proper `_device`  attribute, but for the device itself, this'd cause a circular reference, making it harder for the gc to clean up.

Reply:
mmm, good point ;)

Reply:
Given that device already has a loop and we haven't noticed any problems, I think it would be cleaner if device._device = self.  

I'm happy to undo this last change if you agree.  Either way, I'll abide by your decision.

Reply:
whatever happened to "weak references" i never learned how to use them....

Reply:
but i always thought they were meant for these usecases.

Reply:
For a language like Swift that only has reference counting, this would be a perfect use case for a weak reference. In a language like Python which already has a cyclic garbage collector, it's less necessary.  A live reference to the queue will keep the device from being GCed, and a live reference to the device will keep the queue from being GC'ed. When there's no live reference to either of them, both will be collected when the GC runs.

If it were absolutely critical that devices be GCed as quickly as possible, weak references might make sense.  But no one even noticed until I pointed it out. In fifteen years of Python programming, I've hardly ever seen weak references used.

So my comment again to @almarklein. If you think that device._device should be the device, I'll bring back that code.  If you think it should stay as it is now, close this conversation, approve, and merge.

Reply:
I agree. Weak references have value when you want something to have access to a certain object, without it preventing that object to clean up. They feel like overkill in a situation like this. Cyclic references are still cleaned up as long as both are unreachable. It is my understanding that it only can take a bit longer for that to happen.

The device-object is a pretty long-lived object, so the speed of collection does not matter much, and trying to remove both cyclic references will make the code more complex, which is not worth it.

So yeah, let's move on :)

Reply:
>  I think it would be cleaner if device._device = self.

Yes, let's revert the last change.